src/transformStateWithClones.js

```diff
@@ -6,8 +6,53 @@
  *
  * @return {Object[]}
  */
+
+function removeProperties(object, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete object[key];
+  }
+}
+
+function clearObject(object) {
+  for (const key in object) {
+    delete object[key];
+  }
+}
+
+function addProperties(object, propertiesToAdd) {
+  Object.assign(object, propertiesToAdd);
+}
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateClones = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        stateClones[stateClones.length] = { ...stateCopy };
```

Comment:
You can use `stateClones.push({ ...stateCopy })` instead of `stateClones[stateClones.length] = { ...stateCopy }`.

And you can write it only once, right after `switch case` to avoid repeating code